pkg/api/kubegraph/analysis/hpa.go

```diff
@@ -0,0 +1,178 @@
+		node := uncastNode.(*kubenodes.HorizontalPodAutoscalerNode)
+
+		scaledObjects := graph.SuccessorNodesByEdgeKind(
+			uncastNode,
+			kubegraph.ScalingEdgeKind,
+		)
+
+		if len(scaledObjects) < 1 {
+			markers = append(markers, createMissingScaleRefMarker(node, nil, namer))
+		}
+
+		for _, scaleRef := range scaledObjects {
+			if existenceChecker, ok := scaleRef.(osgraph.ExistenceChecker); ok && !existenceChecker.Found() {
+				// if this node is synthetic, we can't be sure that the HPA is scaling something that actually exists
+				markers = append(markers, createMissingScaleRefMarker(node, scaleRef, namer))
+			}
+		}
+	}
+
+	return markers
+}
+
+func createMissingScaleRefMarker(hpaNode *kubenodes.HorizontalPodAutoscalerNode, scaleRef graphapi.Node, namer osgraph.Namer) osgraph.Marker {
+	return osgraph.Marker{
+		Node:         hpaNode,
+		Severity:     osgraph.ErrorSeverity,
+		RelatedNodes: []graphapi.Node{scaleRef},
+		Key:          HPAMissingScaleRefError,
+		Message:      fmt.Sprintf("%s is attempting to scale %s/%s, which doesn't exist", namer.ResourceName(hpaNode), hpaNode.Spec.ScaleRef.Kind, hpaNode.Spec.ScaleRef.Name),
+	}
+}
+
+// FindOverlappingHPAs scans the graph in search of HorizontalPodAutoscalers that are attempting to scale the same set of pods.
+// This can occur in two ways:
+//   - 1. label selectors for two ReplicationControllers/DeploymentConfigs/etc overlap
+//   - 2. multiple HorizontalPodAutoscalers are attempting to scale the same ReplicationController/DeploymentConfig/etc
+// Case 1 is handled by deconflicting the area of influence of ReplicationControllers/DeploymentConfigs/etc, and therefore we
+// can assume that it will be handled before this step. Therefore, we are only concerned with finding HPAs that are trying to
+// scale the same resources.
+//
+// The algorithm that is used to implement this check is decribed as follows:
+//  - create a sub-graph containing only HPA nodes and other nodes that can be scaled, as well as any scaling edges or other
+//    edges used to connect between objects that can be scaled
+//  - for every resulting edge in the new sub-graph, create an edge in the reverse direction
+//  - find the shortest paths between all HPA nodes in the graph
+//  - shortest paths connecting two horizontal pod autoscalers are used to create markers for the graph
+func FindOverlappingHPAs(graph osgraph.Graph, namer osgraph.Namer) []osgraph.Marker {
+	markers := []osgraph.Marker{}
+
+	nodeFilterFunction := osgraph.NodesOfKind(
+		kubenodes.HorizontalPodAutoscalerNodeKind,
+		kubenodes.ReplicationControllerNodeKind,
+		deploynodes.DeploymentConfigNodeKind,
+	)
+	edgeFilterFunction := osgraph.EdgesOfKind(
+		kubegraph.ScalingEdgeKind,
+		deploygraph.DeploymentEdgeKind,
+	)
+
+	hpaSubGraph := graph.Subgraph(nodeFilterFunction, edgeFilterFunction)
+	for _, edge := range hpaSubGraph.Edges() {
+		osgraph.AddReversedEdge(hpaSubGraph, edge.From(), edge.To())
```

Comment:
+1 on making `AddReversedEdge` accept just the edge. @deads2k I don't see why @stevekuznetsov shouldn't rework it in this PR.